tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/input_inline_fusion_pass.cc

```diff
@@ -0,0 +1,390 @@
+                                 Operation* producer, LoadOp load_op,
+                                 std::vector<Operation*> load_ops) const;
+  bool checkIfFusible(Operation* user, Operation* producer, LoadOp load_op,
+                      bool& can_remove_producer,
+                      std::vector<Operation*>& load_ops) const;
+};
+
+Operation* InputInlineFusionPattern::getFusibleOperation(LoadOp load_op) const {
+  Operation* lhlo_op = nullptr;
+  for (auto* user : load_op.getMemRef().getUsers()) {
+    if (isa<LmhloOp>(user) && (cast<LmhloOp>(user).getResultBuffer() ==
+                               load_op.getOperation()->getOperand(0))) {
+      if (lhlo_op == nullptr) {
+        lhlo_op = user;
+      } else {
+        assert(false &&
+               "More than one lhlo_op write to one Memref within one fusion");
+      }
+    }
+  }
+  return lhlo_op;
+}
+
+// Check if there are no other consumers of the producer
+// except the ParallelOp.
+bool InputInlineFusionPattern::checkIfFusible(
+    Operation* user, Operation* producer, LoadOp load_op,
+    bool& can_remove_producer, std::vector<Operation*>& load_ops) const {
+  assert(isa<scf::ParallelOp>(user) &&
+         "User is expected to be scf::ParallelOp");
+  load_ops.clear();
+  assert(isa<LmhloOp>(producer) && "Unexpected producer in checkIfFusible");
+  auto producer_result_memref = cast<LmhloOp>(producer).getResultBuffer();
+  can_remove_producer = true;
+  auto lhlo_dialect = user->getContext()->getLoadedDialect("lmhlo");
+  for (auto* memref_user : producer_result_memref.getUsers()) {
+    if ((memref_user->getDialect() == lhlo_dialect) &&
+        (memref_user != producer)) {
+      return false;
+    } else if (isa<LoadOp>(memref_user)) {
+      LoadOp other = cast<LoadOp>(memref_user);
+      if (other.getMemRef() == load_op.getMemRef() &&
+          other.getIndices() == load_op.getIndices()) {
+        // llvm::to_vector<4>(other.getIndices()) ==
+        // llvm::to_vector<4>(load_op.getIndices())) {
+        load_ops.emplace_back(memref_user);
+      } else {
+        can_remove_producer = false;
+      }
+      // TODO:: check the memref_user is inside the loops
+    }
+  }
+  return true;
+}
+
+template <typename LHLO_OpTy>
+bool elemwiseFuseHelper(PatternRewriter& rewriter, Operation* user,
+                        Operation* producer, LoadOp load_op,
+                        std::vector<Operation*>& load_ops) {
+  if (!isa<LHLO_OpTy>(producer)) {
+    return false;
+  }
+  auto loc = user->getLoc();
+  SmallVector<Value, 4> operand_values;
+  unsigned num_operands = producer->getNumOperands();
+  for (unsigned i = 0; i < num_operands - 1; ++i) {
+    auto producer_operand = producer->getOperand(i);
+    rewriter.setInsertionPoint(load_op);
+    operand_values.push_back(
+        rewriter.create<LoadOp>(loc, producer_operand, load_op.getIndices()));
+  }
+  auto inlined_result =
+      HloOpToStdScalarOp::map<LHLO_OpTy>(llvm::cast<LHLO_OpTy>(producer),
```

Comment:
Can we make this mapping runtime? This way we can remove the templates.

Reply:
Not sure if i get the point. HloOpToStdScalarOp::map<LHLO_OpTy> is already previously used a several existing passes. Here it's just to reuse the codes.

Reply:
By "runtime" I mean not using C++ templates, e.g. `HloOpToStdScalarOp::map(mlir::Operation* producer)` instead. I believe that `mlir::Operation::getName()` and `mlir::Operation::getDialect()` give us enough information to do the mapping.

Reply:
The map is from hlo -> std, more things are to be considered besides of getName(). 
For example hlo.div may be mapped to std.divf, std.UnsignedDivIOp or std.SignedDivIOp, thus map_lmhlo_to_scalar_op.h is needed.

Reply:
Sure, still we can do mappings at C++ runtime: op.getName() + op.getType() + ... -> the corresponding std op.

The point is to not use C++ templates. As you notice using C++ templates also requires to maintain a closed collection of all possible ops, which is painful to maintain and incompatible with the existing elementwise trait.